### PR TITLE
feat/add_room_database

### PR DESCRIPTION
**Pull request by boe-dev**

### CODE REVIEW

File: app/src/main/java/dev/five_star/mybooks/database/BookDao.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+interface BookDao {
+
+    @Query("SELECT * FROM book_table")
+    fun getAllBooks(): Flow<List<Book>>
+
+    @Query("SELECT * FROM book_table WHERE id = :bookId")
+    fun getBook(bookId: Int) : Book
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertBook(book: Book)
```
Comment (n8ebel) [inline app/src/main/java/dev/five_star/mybooks/database/BookDao.kt:19]: I think we mentioned this last time, but I'll mention again just to be thorough.  There's an `@Update` annotation that we can use if/when needed.  So if you wanted, you could make your apis on this DAO more explicit by separating out the insert and update functionality into 2 separate methods.

File: app/src/main/java/dev/five_star/mybooks/booklist/MainViewModel.kt [hunk 2]
```diff
@@ -28,34 +33,44 @@ class MainViewModel(private var repository: BookRepository) : ViewModel() {
...
+
+                repository.addBook(newBook)
                 _dialogEffect.postValue(DialogEffect.CloseAddBook)
-                getBookList()
+
+                //TODO how can I do this with room?
+//                val entered = repository.addBook(newBook)
+//                if (entered) {
+//                    _dialogEffect.postValue(DialogEffect.CloseAddBook)
+//                }
```
Comment (n8ebel) [inline app/src/main/java/dev/five_star/mybooks/booklist/MainViewModel.kt:65]: Let's chat through this bit

File: app/src/main/java/dev/five_star/mybooks/book_details/DetailsViewModel.kt [hunk 3]
```diff
@@ -52,7 +61,8 @@ class DetailsViewModel(private val book: Book, private val repository: BookRepos
             repository.addPageEntry(enteredPage)
             //TODO I think this should not happen directly on the view
             pageEntry.postValue(null)
-            _pagesList.postValue(repository.getPagesForBook(book))
+//           TODO: how to replace this?
+//            _pagesList.postValue(repository.getPagesForBook(book))
```
Comment (n8ebel) [inline app/src/main/java/dev/five_star/mybooks/book_details/DetailsViewModel.kt:65]: Let's chat through this together

**Review by n8ebel — COMMENTED** [review]
Nice work 👍🏼     Let's walk through this PR/comments during our session today and make sure we touch on some of the questions you left in your code.

File: app/src/main/java/dev/five_star/mybooks/book_details/DetailsFragment.kt [hunk 4]
```diff
@@ -21,8 +21,9 @@ class DetailsFragment : Fragment() {
     private val pagesEntryAdapter = PagesAdapter()
     private val args: DetailsFragmentArgs by navArgs()
 
+
     private val viewModel: DetailsViewModel by viewModels {
-        DetailsViewModelFactory(args.book, BookRepository)
+        DetailsViewModelFactory(args.bookId, (requireActivity().application as MyBookApplication).bookRepository)
```
[the remark this answers is not in the document]
Comment (boe-dev) [reply app/src/main/java/dev/five_star/mybooks/book_details/DetailsFragment.kt:26]: that's a nice idea, I putted them directly into the MyBookApplication.kt

**Pull request merged**